src/vm/database/sqlite.rs

```diff
@@ -0,0 +1,305 @@
+        Ok(contract_db)
+    }
+
+    fn check_schema(&self) -> Result<()> {
+        let sql = "SELECT sql FROM sqlite_master WHERE name=?";
+        if let Some(ref conn) = self.conn {
+            let _: String = conn.query_row(sql, &["maps_table"],
+                                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            let _: String = conn.query_row(sql, &["contracts"],
+                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            let _: String = conn.query_row(sql, &["data_table"],
+                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            Ok(())
+        } else {
+            Err(Error::new(ErrType::SqlConnectionClosed))
+        }
+    }
+
+    fn inner_open(filename: &str) -> Result<SqliteContractDatabase> {
+        let conn = Connection::open(filename)
+            .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+        Ok(SqliteContractDatabase {
+            conn: Some(conn),
+            save_point: 0
+        })
+    }
+
+    fn load_map(&self, contract_name: &str, map_name: &str) -> Result<SqliteDataMap> {
+        let (map_identifier, key_type, value_type): (_, String, String) =
+            self.query_row(
+                "SELECT map_identifier, key_type, value_type FROM maps_table WHERE contract_name = ? AND map_name = ?",
+                &[contract_name, map_name],
+                |row| {
+                    (row.get(0), row.get(1), row.get(2))
+                })?
+            .ok_or(Error::new(ErrType::UndefinedMap(map_name.to_string())))?;
+
+        Ok(SqliteDataMap {
+            map_identifier: map_identifier,
+            key_type: TypeSignature::deserialize(&key_type)?,
+            value_type: TypeSignature::deserialize(&value_type)?
+        })
+    }
+
+    fn load_contract(&self, contract_name: &str) -> Result<Option<Contract>> {
+        let contract: Option<String> =
+            self.query_row(
+                "SELECT contract_data FROM contracts WHERE contract_name = ?",
+                &[contract_name],
+                |row| {
+                    row.get(0)
+                })?;
+        match contract {
+            None => Ok(None),
+            Some(ref contract) => Ok(Some(Contract::deserialize(contract)?))
+        }
+    }
+
+    fn execute<P>(&mut self, sql: &str, params: P) -> Result<usize>
+    where
+        P: IntoIterator,
+        P::Item: ToSql {
+        match self.conn.take() {
+            Some(conn) => {
+                let result = conn.execute(sql, params)
+                    .map_err(|x| {
+                        eprintln!("SQL Execution Error: {:?}", x);
+                        Error::new(ErrType::SqliteError(IncomparableError{ err: x }))
```

Comment:
I'm not sure what information the `IncomparableError` type is supposed to be conveying.  Is it expected that some SQL commands will fail to execute in the node's normal operation?  If so, are all such failures due to something not being comparable?

Reply:
No, the IncomparableError is an implementation weirdness. We want the SqliteError type to refer to the originating RusqliteError, but also, for the purposes of testing, it's useful that the error class implement `PartialEq`. Wrapping the rusqlite error in that struct allowed me to use `derive` for the PartialEq, rather than implementing equality checks directly on the enum.

Reply:
Ah, you're saying that `IncomparableError` error variant turns an otherwise non-`PartialEq`-implementing error into something for which `PartialEq` is defined?

Reply:
Yep, exactly. The IncomparableError just always returns false for the equality check.